mautrix_signal/portal.py

```diff
@@ -1237,6 +1237,68 @@ async def handle_signal_delete(self, sender: p.Puppet, message_ts: int) -> None:
         except MForbidden:
             await self.main_intent.redact(message.mx_room, message.mxid)
 
+    # endregion
+    # region Matrix -> Signal metadata
+
+    async def create_signal_group(
+        self, source: u.User, levels: PowerLevelStateEventContent
+    ) -> None:
+        user_mxids = await self.az.intent.get_room_members(
+            self.mxid, (Membership.JOIN, Membership.INVITE)
+        )
+        invitee_addresses = []
+        for mxid in user_mxids:
+            mx_user = await u.User.get_by_mxid(mxid, create=False)
+            if mx_user and mx_user.address and mx_user.username != source.username:
+                invitee_addresses.append(mx_user.address)
+            puppet = await p.Puppet.get_by_mxid(mxid, create=False)
+            if puppet:
+                invitee_addresses.append(puppet.address)
+        avatar_path: str | None = None
+        if self.avatar_url:
+            avatar_data = await self.az.intent.download_media(self.avatar_url)
+            self.avatar_hash = hashlib.sha256(avatar_data).hexdigest()
+            avatar_path = self._write_outgoing_file(avatar_data)
+            self.avatar_set = True
+        signal_chat = await self.signal.create_group(
+            source.username, title=self.name, members=invitee_addresses, avatar_path=avatar_path
+        )
+        if avatar_path and self.config["signal.remove_file_after_handling"]:
+            try:
+                os.remove(avatar_path)
+            except FileNotFoundError:
+                pass
+        self.chat_id = signal_chat.id
+        await self.signal.update_group(
+            username=source.username,
+            group_id=self.chat_id,
+            update_access_control=GroupAccessControl(
+                members=(
+                    AccessControlMode.MEMBER
+                    if levels.invite == 0
+                    else AccessControlMode.ADMINISTRATOR
+                ),
+                attributes=None,
+                link=None,
+            ),
+        )
+        await self.signal.update_group(
+            username=source.username,
+            group_id=self.chat_id,
+            update_access_control=GroupAccessControl(
+                attributes=(
+                    AccessControlMode.MEMBER
+                    if levels.state_default == 0
+                    else AccessControlMode.ADMINISTRATOR
+                ),
+                members=None,
+                link=None,
+            ),
+        )
+        await self._postinit()
+        await self.insert()
```

Comment:
These should probably be immediately after the `create_group` call (it might also need some locking to make sure another portal isn't created)